opm/parser/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -766,6 +773,36 @@ namespace Opm {
 
         return index(i,j,k,c);
     }
+
+    void ZcornMapper::fixupZCORN( std::vector<double>& zcorn, int sign, size_t i, size_t j, size_t k , size_t c) {
+        size_t index1 = this->index(i,j,k,c);
+        size_t index2 = this->index(i,j,k,c+4);
+
+        if ((zcorn[index2] - zcorn[index1]) * sign < 0 )
+            zcorn[index2] = zcorn[index1];
+
+    }
+
+    void ZcornMapper::fixupZCORN( std::vector<double>& zcorn) {
+        int sign = zcorn[ this->index(0,0,0,0) ] < zcorn[this->index(0,0, this->dims[2] - 1,0)] ? 1 : -1;
+
+        for (size_t k=0; k < this->dims[2]; k++)
+            for (size_t j=0; j < this->dims[1]; j++)
+                for (size_t i=0; i < this->dims[0]; i++) {
+                    this->fixupZCORN( zcorn , sign, i , j , k , 0 );
+
+                    if (i == (this->dims[0] - 1))
```

Comment:
Yes - you are right. The pillars, i.e. `COORD` are shared between cells, but not the z values.